File: neetcode-loadmap/001-arrays-hashing/005.347.top_k_frequent_elements.py
class Solution:
    def top_k_frequent_elements(self, nums: List[int], k: int) -> List[int]:
        count = {}
        freq = [[] for i in range(len(nums) + 1)]

        for n in nums:
            count[n] = 1+ count.get(n, 0)
        for n, c in count.items():
            freq[c].append(n)
        res = []
        # ignore 0 occurrance
        for i in range(len(freq) -1 , 0 , -1):
            for n in freq[i]:
                res.append(n)
                if len(res) == k:
                    return res

File: neetcode-loadmap/001-arrays-hashing/005.347.top-k-frequent-elements.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function (nums, k) {
  if (nums.length === 1) return nums;
  let count = {};
  let freq = Array.from({ length: nums.length + 1 }, () => []); // from 0 to nums.length; 3개면 0,1,2,3 총 4개 필요
  for (const n of nums) {
    // keys are each number, values are counts
    count[n] = 1 + (count[n] ? count[n] : 0);
  }

  for (const c in count) {
    freq[count[c]].push(c);
  }
  let res = [];
  // 0개는 무시해도 됨
  for (let i = freq.length - 1; i > 0; i--) {
    for (const n of freq[i]) {
      res.push(n);
      if (res.length === k) {
        return res;
      }
    }
  }
};

console.log(topKFrequent([4, 1, -1, 2, -1, 2, 3], 2));
